sound/soc/sof/intel/cnl.c

```diff
@@ -29,6 +30,65 @@ static const struct snd_sof_debugfs_map cnl_dsp_debugfs[] = {
 static void cnl_ipc_host_done(struct snd_sof_dev *sdev);
 static void cnl_ipc_dsp_done(struct snd_sof_dev *sdev);
 
+irqreturn_t cnl_ipc4_irq_thread(int irq, void *context)
+{
+	struct snd_sof_dev *sdev = context;
+	bool ipc_irq = false;
+	u32 hipcida;
+	u32 hipctdr;
+	u32 hipctdd;
+
+	hipcida = snd_sof_dsp_read(sdev, HDA_DSP_BAR, CNL_DSP_REG_HIPCIDA);
+	hipctdr = snd_sof_dsp_read(sdev, HDA_DSP_BAR, CNL_DSP_REG_HIPCTDR);
+	hipctdd = snd_sof_dsp_read(sdev, HDA_DSP_BAR, CNL_DSP_REG_HIPCTDD);
+
+	if (hipcida & CNL_DSP_REG_HIPCIDA_DONE)
+		/* mask Done interrupt */
+		snd_sof_dsp_update_bits(sdev, HDA_DSP_BAR, CNL_DSP_REG_HIPCCTL,
+					CNL_DSP_REG_HIPCCTL_DONE, 0);
+
+	/* new message from DSP */
+	if (hipctdr & CNL_DSP_REG_HIPCTDR_BUSY) {
+		u32 msg = hipctdr & CNL_DSP_REG_HIPCTDR_MSG_MASK;
+		u32 msg_ext = hipctdd & CNL_DSP_REG_HIPCTDD_MSG_MASK;
+
+		if (hipctdr & SOF_IPC4_MSG_DIR_MASK) {
+			struct sof_ipc4_msg *data = sdev->ipc->msg.reply_data;
+
+			data->primary = msg;
+			data->extension = msg_ext;
+
+			spin_lock_irq(&sdev->ipc_lock);
+
+			snd_sof_ipc_get_reply(sdev);
+			snd_sof_ipc_reply(sdev, data->primary);
```

Comment:
where do you process reply status ?  It is not correct to process reply message here.  FW send another ipc message for the status of last message. It should be process in  else path, not here

Reply:
you can send a error ipc message to validate this PR

Reply:
ok, I understand your code

Reply:
@RanderWang, during audio playback I can use the msg inject.
Sending PAUSE for the pipeline (0x13010003) - supported, success:
```
[  132.806420] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx      : 0x13010003|0x0: GLB_SET_PIPELINE_STATE
[  132.807194] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx reply: 0x33000000|0x0: GLB_SET_PIPELINE_STATE
[  132.807211] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx done : 0x13010003|0x0: GLB_SET_PIPELINE_STATE
```

Sending RESET to the pipeline (0x13010002) - not supported, fail:
```
[  655.677763] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx      : 0x13010002|0x0: GLB_SET_PIPELINE_STATE
[  655.678219] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx reply: 0x33000006|0x1: GLB_SET_PIPELINE_STATE
[  655.678226] sof-audio-pci-intel-tgl 0000:00:1f.3: FW reported error: 6 - Unknown error while processing the request
[  655.678241] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc error for msg 0x13010002|0x0
[  655.678246] snd_sof_ipc_msg_injector snd_sof.msg_injector.0: IPC message send failed: -22
```
Do we need to read something else after this?